### ISSUE TITLE
Checkpoint WAL in Head only once in a DB.Compact() call

### ISSUE THREAD

**Issue by codesome** [prometheus/prometheus#7574]
It is possible that Head compaction can fail continuously for various reasons (disk issues, disk capacity, unsolved corruptions in the head, etc), in such cases a lot of WAL would be piled up until the issue is fixed.

When Prometheus starts finally, let's say we have 24h of data in Head not compacted, this will lead to 11 Head compactions (22h of data) in total one after another (in a single `DB.Compact()` call).

Currently, for each compaction, we truncate the Head, which also results into creating a checkpoint. And as its 24h now, the checkpoint is huge (2/3 of 24h data). With 11 compactions like this, it is a lot of re-write of the same data again and again into the checkpoint and the entire process becomes _very_ slow.

So I am proposing to truncate the Head only once in a `DB.Compact()` method so that in the example above, it would lead to a single checkpoint creation.

One of the issues with this is Prometheus coming down in between the compaction process and hence the truncation not happening. So with this proposal, additionally, we could check for possible truncation on startup. (Update: Created https://github.com/prometheus/prometheus/issues/7575 for it)

**Comment by bwplotka** [comment]
cc @kakkoyun as we investigate some issue near this place.

**Comment by kakkoyun** [comment]
> One of the issues with this is Prometheus coming down in between the compaction process and hence the truncation not happening. 

@codesome Seems like we've been hit by this last part in Thanos Receive. We hit this case when a shutdown process terminated prematurely, where we compact the head block. As a result, a block created but the head is not truncated. We ended up creating overlapping blocks. It'd be really nice to have an additional mechanism to prevent it.

**Comment by codesome** [comment]
Opened https://github.com/prometheus/prometheus/issues/7575 dedicated to truncating WAL on startup

**Comment by Harkishen-Singh** [comment]
Will try to take up this.

**Comment by codesome** [comment]
There is already https://github.com/prometheus/prometheus/pull/7579, maybe you could review it?
